packages/plugin-core/src/components/doctor/utils.ts

```diff
@@ -26,6 +26,10 @@ export class DoctorUtils {
     const { vaults, wsRoot, engine } = extension.getDWorkspace();
     let vault;
     try {
+      if ((await vscode.workspace.fs.stat(document.uri)).ctime === 0) {
```

Comment:
Would have preferred to instead use something like this in L24:
```suggestion
      if(!NoteUtils.isNote(document.uri)) return;
```

Also I could not find a good alternative to `workspace.fs.stat()` in the vscode API.

Reply:
I believe `vscode.workspace.fs.stat(document.uri)` will throw a `FileNotFound` error (see: [vscode.d.ts](https://github.com/microsoft/vscode/blob/6a62d4fbdc900506420da1aa8becf3a6354a1c1e/src/vscode-dts/vscode.d.ts#L8192)) so the if block won't be necessary.

> but it would have been outside the plugin-core package and require to either pass a workspace reference or use node fs.

A few line above we already have a check that's exactly what `NoteUtils.isNote` is doing, so I don't think this is sufficient.

For deleted notes that are still open in the editor, `document.uri.fsPath` will still end with `.md`, it's just that there isn't a file there anymore. 

So your solution of trying `vscode.workspace.fs.stat(document.uri)` and just returning when it throws would do the trick.

Reply:
Hey, no worries. Thanks for the response.

#### Regular Editor View
> I believe `vscode.workspace.fs.stat(document.uri)` will throw a `FileNotFound` error (see: [vscode.d.ts](https://github.com/microsoft/vscode/blob/6a62d4fbdc900506420da1aa8becf3a6354a1c1e/src/vscode-dts/vscode.d.ts#L8192)) so the if block won't be necessary.

Good point. Will change it to one line. It covers all deleted files in the regular view. 


#### Source control view
The above line does not work in Source Control view, which is why I checked for creation time:

```js
// existing
{
  ctime: 1666793183976,
  mtime: 1666793183976,
  size: 120,
  type: 1,
};

// deleted, but in source control view
{
  ctime: 0,
  mtime: 1667414348797,
  size: 0,
  permissions: undefined,
  type: 1,
};
```

But it is a bit hacky and I think I will check for uri scheme instead:

```js
// example document.uri in default view
{
  scheme: "file",
  authority: "",
  path: "/z:/Projekte/dev/dendron-example/notes/some.note.md",
  query: "",
  fragment: "",
  _formatted: "file:///z%3A/Projekte/dev/dendron-example/notes/some.note.md",
  _fsPath: "z:\\Projekte\\dev\\dendron-example\\notes\\some.note.md",
}

// example document.uri in source control view
{
  scheme: "git",
  authority: "",
  path: "/z:/Projekte/dev/dendron-example/notes/some.note.md",
  query: "{\"path\":\"z:\\\\Projekte\\\\dev\\\\dendron-example\\\\notes\\\\some.note.md\",\"ref\":\"HEAD\"}",
  fragment: "",
  _formatted: "git:/z%3A/Projekte/dev/dendron-example/notes/some.note.md?%7B%22path%22%3A%22z%3A%5C%5CProjekte%5C%5Cdev%5C%5Cdendron-example%5C%5Cnotes%5C%5Csome.note.md%22%2C%22ref%22%3A%22HEAD%22%7D",
  _fsPath: "z:\\Projekte\\dev\\dendron-example\\notes\\some.note.md",
}
```

**Question:**
Do you think there will be any problems, if I tell DoctorUtils to skip files with `git` uri scheme?
`if (document.uri.scheme === "git") return;`

Pros would be:
- no more frontmatter check on changed files in source control
- no more "duplicate id" false positives when in source control

Reply:
Sorry, I think I will change or cancel my PR.

I created a test in `DuplicateNote.test.ts` and noticed that the test result is the same with and without the code changes. After some reflection what I am basically doing is just avoiding 2 calls + error message when it is a deleted note, by adding 1 unnecessary `fs` call to everything.

What I would really like to do though is to skip files in source control view, as mentioned before. It works, but I do not know how to test for that.

---

<details>
<summary>Here is my current test for deleted files</summary>

```typescript
describeSingleWS(
    "GIVEN an open file that has been deleted",
    {
      postSetupHook: ENGINE_HOOKS.setupEmpty,
    },
    () => {
      test("THEN do nothing", async () => {
        const { engine, wsRoot, vaults } = ExtensionProvider.getDWorkspace();

        await NoteTestUtilsV4.createNoteWithEngine({
          fname: "deleted",
          vault: vaults[0],
          wsRoot,
          body: "note will be deleted",
          engine,
        });

        const vaultPath = VaultUtils.getRelPath(vaults[0]);
        const deletedFilePath = path.join(wsRoot, vaultPath, "deleted.md");
        const deletedFileUri = vscode.Uri.file(deletedFilePath);
        await VSCodeUtils.openFileInEditor(deletedFileUri);

        fs.unlinkSync(deletedFilePath);

        await VSCodeUtils.openFileInEditor(deletedFileUri);
        const editor = VSCodeUtils.getActiveTextEditor();
        const document = editor?.document;

        const resp = await DoctorUtils.findDuplicateNoteFromDocument(document!);
        expect(resp).toEqual(undefined);
      });
    }
  );
```

</details>

Reply:
I think the test case needs a duplicate note to see the effect. otherwise in both cases have no duplicate and the doctor util will do nothing.

Reply:
> Do you think there will be any problems, if I tell DoctorUtils to skip files with git uri scheme?
> if (document.uri.scheme === "git") return;
> Pros would be:
>
>    no more frontmatter check on changed files in source control
>    no more "duplicate id" false positives when in source control

Can you elaborate? is this the editor that opens when you click on a file from the source control side panel?
To be safe with this, we would have to know if regular editors never have git scheme urls. If that is the case I say that would be fine. I can't recall anything regarding the document uri being git scheme in the regular text editor but if you can find info on this matter that would be awesome.

Reply:
## Source Control View

> Can you elaborate? is this the editor that opens when you click on a file from the source control side panel?
To be safe with this, we would have to know if regular editors never have git scheme urls. If that is the case I say that would be fine. I can't recall anything regarding the document uri being git scheme in the regular text editor but if you can find info on this matter that would be awesome.

Yeah, that's the one. I could only find out so far that an extension can implement its own scheme and the git scheme is being used by the builtin vscode git extension.

- [StackOverflow: gitURI scheme](https://stackoverflow.com/questions/56383584/whats-the-git-uri-scheme-and-is-it-documented-anywhere/71556348#71556348)
- [TextDocumentContentProvider](https://code.visualstudio.com/api/extension-guides/virtual-documents#textdocumentcontentprovider)

## Deleted Notes

> I think the test case needs a duplicate note to see the effect. otherwise in both cases have no duplicate and the doctor util will do nothing.

I created the test but as I mentioned: Dendron already checks for deleted files [a few lines later](https://github.com/dendronhq/dendron/blob/78c5b39c94a2fc3bc5b5f9ce111009b737663914/packages/plugin-core/src/components/doctor/utils.ts#L41-L46).

```typescript
    const resp = file2Note(fsPath, vault);
    if (ErrorUtils.isErrorResp(resp)) {
      // not in file system, we do nothing.
      Logger.error({ ctx, error: resp.error });
      return;
    }
```

<details>
<summary>Current Test</summary>

```typescript
  describeSingleWS(
    "GIVEN an open duplicate note that has been deleted",
    {
      postSetupHook: ENGINE_HOOKS.setupEmpty,
    },
    () => {
      test("THEN do nothing", async () => {
        const { engine, wsRoot, vaults } = ExtensionProvider.getDWorkspace();
        const vaultPath = VaultUtils.getRelPath(vaults[0]);

        await NoteTestUtilsV4.createNoteWithEngine({
          fname: "deleted",
          vault: vaults[0],
          wsRoot,
          body: "note will be deleted",
          engine,
        });
        const delNotePath = path.join(wsRoot, vaultPath, "deleted.md");
        const delNoteUri = vscode.Uri.file(delNotePath);
        const dupeNotePath = path.join(wsRoot, vaultPath, "dupe.md");
        const noteContent = fs.readFileSync(delNotePath, { encoding: "utf-8" });
        fs.writeFileSync(dupeNotePath, noteContent, { encoding: "utf-8" });

        await VSCodeUtils.openFileInEditor(delNoteUri);
        fs.unlinkSync(delNotePath);

        const editor = VSCodeUtils.getActiveTextEditor();
        const document = editor?.document;

        const resp = await DoctorUtils.findDuplicateNoteFromDocument(document!);
        expect(resp).toEqual(undefined);
      });
    }
  );
```

</details>

### Working as intended

When I first wrote the issue I thought the note was supposed to be ignored, like a file without frontmatter. The error message popup made it look to me like something broke and required action:

![grafik](https://user-images.githubusercontent.com/8905319/200140810-38c3bd36-57c5-4a81-a93a-839092858401.png)

But now in the code I see that it is working as intended. So my change would only prevent the error popup. If that is a desired change, then I can do it with the oneliner you suggested:

```typescript
    // return if file is not a markdown
    if (!fsPath.endsWith(".md")) return;
    // return if file is in source control view
    if (document.uri.scheme === "git") return;

    const extension = ExtensionProvider.getExtension();
    const { vaults, wsRoot, engine } = extension.getDWorkspace();
    let vault;
    try {
      await vscode.workspace.fs.stat(document.uri); // Skip if file does not exist
      vault = VaultUtils.getVaultByFilePath({
        vaults,
        wsRoot,
        fsPath,
      });
    } catch (error: any) {
      // document doesn't exist in workspace.
      return;
    }
```